src/backend/distributed/transaction/transaction_management.c

```diff
@@ -691,3 +693,17 @@ IsMultiStatementTransaction(void)
 		return false;
 	}
 }
+
+
+/*
+ * ExecutingUDF returns true if we are executing a function call.
+ */
+static bool
+ExecutingUDF(void)
+{
+	/*
+	 * SQL UDFs start another executor, and PL UDFs use SPI to
+	 * execute statements.
+	 */
+	return ExecutorLevel > 1 || SPILevel > 0;
```

Comment:
This probably also triggers for other things other than UDFs (e.g. recursively planned CTEs).

Reply:
Renamed the function to `MaybeExecutingUDF`, I guess in other cases we also create remote transaction blocks too, so false positives here should be safe. I need to do some testing before  merging though.